src/corpora/auto.py

```diff
@@ -29,20 +34,41 @@ def get_auto_dataset(
     preprocessing_num_proc: int = 64,
     stride: int = -1,
     ignore_train: bool = False,
+    streaming: bool = True,
 ) -> datasets.DatasetDict:
     """Run basic tokenization and grouping to turn a Hugging Face Dataset (via `datasets`) into a torch.Dataset."""
 
     # Sanity check on input args
     stride = seq_len if stride < 0 else stride
     assert stride <= seq_len, f"Data grouping stride ({stride}) is smaller than sequence length: we are losing data."
+
     dataset = datasets.load_dataset(
-        dataset_id, name=dataset_name, cache_dir=str(paths["dataset"]), keep_in_memory=True
+        dataset_id, name=dataset_name, cache_dir=str(paths["dataset"]), keep_in_memory=not streaming,
+        streaming=streaming,
     )
 
+    # while HF datasets has a map method for both normal Dataset and streaming IterableDataset, the IterableDataset
+    # version doesn't actually work with Trainer so we have to wrap it in this StreamedDataset nonsense.
+    # The reasons are:
+    # 1. IterableDataset transforms are pickled using pickle (not dill or something reasonable), so you have to be super
+    # careful about what you pass to map. This is mostly Python's fault.
+    # 2. IterableDataset doesn't have a __len__ method, which is required by Trainer for non-torch.utils.IterableDataset
+    # types. You should be able to fix this by calling dataset.with_format(type="torch") but that's not working for
+    # reason #1 (they use a local class which can't be pickled). You can work around this too (see ensure_format).
+    # 3. IterableDataset, when successfully pickled, tries to `open` the urls of the dataset as files (rather than
+    # downloading them), which of course explodes. I haven't figured out how to work around this yet.
+    if streaming:
+        detokenizer = DATASET_TOKENIZATION_REGISTRY.get(dataset_id, None)
+        return IterableDatasetDict({
+            k: StreamedDataset(dataset_id, dataset_name, k, str(paths["dataset"]), tokenizer, detokenizer, seq_len,
+                               stride) for k, _ in dataset.items()
+        })
+
     if "validation" not in dataset:
         assert "train" in dataset, "You must have train in dataset to make a validation dataset"
         # Create Dataset Split Cache Files
         train_fn, val_fn = [str(paths["dataset"] / dataset_id / f"{k}-split.hf") for k in ["train", "val"]]
+        # TODO: what to do for streaming?
```

Comment:
Probably raise an error here -- if streaming without a fixed validation set, either set a `no_validation` flag, or raise hell.

Reply:
yeah that's probably the right thing